README.md
# flask-eventics

This is a [Flask Blueprint](http://flask.pocoo.org/docs/0.10/blueprints/) for generating [iCalendar files](https://www.ietf.org/rfc/rfc2445.txt) from JSON returned by a REST API Endpoint.

**Status:** Proof-of-concept.

## Requirements

Requirements can be satisfied with `pip`:

```shell
$ pip install -r requirements.txt
```

* [Flask](http://flask.pocoo.org/)
* [Python icalendar](http://icalendar.readthedocs.org/en/latest/) 
* [Python Dateutil](https://dateutil.readthedocs.org/en/latest/)
* [Python mock (for the tests)](http://www.voidspace.org.uk/python/mock/)

## Installation

To clone and install flask-eventics locally in an existing Python 
[`virtualenv`](https://virtualenv.pypa.io/en/latest/):

```shell
$ git clone https://github.com/cfpb/flask-eventics
$ pip install -e flask-eventics
```

Note: this installs flask-eventics in 'editable' mode.

flask-eventics can be installed directly from Github:

```shell
$ pip install git+https://github.com/cfpb/flask-eventics
```

Once installed, the flask-eventics blueprint simply needs to be
registered with your flask application:

```python
from flask import Flask
from flask_eventics import eventics

app = Flask(__name__)
app.register_blueprint(eventics)
```

If using [Sheer](https://github.com/cfpb/sheer), flask-eventics can be
added to your site's `_settings/blueprints.json` file.

```json
{
  "flask_eventics": {
    "package": "flask_eventics.controllers",
    "module":  "eventics"
  }
}
```

## Configuration

flask-eventics can optionally use the Flask app configuration or
environment variables for configuration. 

#### `EVENT_ICS_URL` 
The URL at which ICS files will be available. It should include a named 
argument, `event_slug`, which will corropsond to an event available 
from a source url.

Example: `/events/<event_slug>/ics`

#### `EVENT_SOURCE` 
The REST URL at which JSON that describes a particular event is available. 
This was designed to work with Elasticsearch, but it should work with any 
REST API that provides all the necessary properties within a single 
JSON object.

Example: `http://localhost:9200/content/event/<event_slug>/_source`

#### `EVENT_CALENDAR_PRODID` 
The iCalendar product identifier that will be returned for generated 
calendars.

#### JSON-iCalendar Field Mapping

There is a series of configuration options that map iCalendar fields to
properties on the JSON object returned by the event source. 

* `EVENT_FIELD_SUMMARY`: `summary`
* `EVENT_FIELD_DTSTART`: `dtstart`
* `EVENT_FIELD_DTEND`: `dtend`
* `EVENT_FIELD_DTSTAMP`: `dtstamp`
* `EVENT_FIELD_UID`: `uid`
* `EVENT_FIELD_PRIORITY`: `priority`
* `EVENT_FIELD_ORGANIZER`: `organizer`
* `EVENT_FIELD_ORGANIZER_ADDR`: `organizer_email`
* `EVENT_FIELD_LOCATION`: `location`
* `EVENT_FIELD_STATUS`: `status`

#### Field Defaults

The following configuration options define defaults that will be used 
if the above fields are not available in the JSON object returned by 
the event source, 

* `EVENT_DEFAULT_SUMMARY`: `''`,
* `EVENT_DEFAULT_DTSTART`: `2015-01-11T10:30:00Z`,
* `EVENT_DEFAULT_DTEND`: `2015-01-11T10:30:00Z`,
* `EVENT_DEFAULT_DTSTAMP`: `2015-01-11T10:30:00Z`,
* `EVENT_DEFAULT_UID`: `''`,
* `EVENT_DEFAULT_PRIORITY`: `1`,
* `EVENT_DEFAULT_ORGANIZER`: `''`,
* `EVENT_DEFAULT_ORGANIZER_ADDR`: `''`,
* `EVENT_DEFAULT_LOCATION`: `''`,
* `EVENT_DEFAULT_STATUS`: `TENTATIVE`,

## Event Sources

An Event Source is a simply a REST API endpoint that returns a JSON
object that describes an event. This could be a custom API or a 
service such as [Elasticsearch](http://www.elasticsearch.org/). With the
default JSON-iCalendar field mappings above the JSON object returned
from the REST endpoint might look like this:

```json
{
    "dtstamp": "2015-01-11T09:29:08Z", 
    "uid": "F964B38DB71F484FA2ABC57F621CB7F1@2015-01-11 09:30:00", 
    "dtend": "2015-01-11T10:30:00Z", 
    "created": "2015-01-11T10:09:42Z", 
    "summary": "A Test Meeting", 
    "location": "Washington, DC", 
    "dtstart": "2015-01-11T09:30:00Z", 
    "id": 24407, 
    "day": "2015-01-11", 
    "description": ""
}
```

This would result in the following iCalendar file being generated:

```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//A Calendar//somewhere.com//
METHOD:publish
BEGIN:VEVENT
SUMMARY:A Test Meeting
DTSTART;VALUE=DATE-TIME:20150111T093000Z
DTEND;VALUE=DATE-TIME:20150111T103000Z
DTSTAMP;VALUE=DATE-TIME:20150111T142908Z
UID:F964B38DB71F484FA2ABC57F621CB7F1@2015-01-11 09:30:00
LOCATION:Washington\, DC. 
ORGANIZER;CN=:MAILTO:
STATUS:TENTATIVE
END:VEVENT
END:VCALENDAR
```

## Licensing 

Public Domain/CC0 1.0

1. [Terms](TERMS.md)
2. [License](LICENSE)
3. [CFPB Source Code Policy](https://github.com/cfpb/source-code-policy/)




CHANGELOG.md

## 0.2.0

### Added
- pytz dependency for timezone use

### Deprecated
- Nothing.

### Removed
- Function to get fields from Event

### Fixed
- Flapjack Event document type wiring.

## 0.1.0 

### Added
- The blueprint to generate ics files from JSON events
- Configuration options via Flask app config or envvars
- Tests for above

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


flask_eventics/tests.py
# -*- coding: utf-8 -*-

import unittest
import mock

import icalendar
# import dateutil

from flask import Flask
from flask_eventics.controllers import eventics, record_state, get_event_json, generate_ics, EVENTICS_CONFIG


class EventICSTestCase(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_record_state(self):
        """
        Test that the application state is appropriately recorded when
        the blueprint is registered. To do this, we'll mock Flask's
        app.config.get and os.environ.get and make sure they're called
        appropriately.

        We want to test our fail-through, wherein:
            If there is a value in app.config, use it
            Otherwise if there is a value in os.environ, us it
            Otherwise use the default value

        We want to configure Flask blueprints differently, or have the
        option to configure them differently, and not just through
        envvars. This is a first step.
        """

        # This is a mock app, with mock config that is simply a dictionary. We
        # don't need to test dictionary lookups, so we'll mock that later. We
        # just need to test that the lookup happens in the right order, the
        # value comes from the appropriate place, and that it gets appropriately
        # assigned.
        mock_app = mock.MagicMock()
        mock_app.config = {}

        # This is a mock state object which gets passed to record_state.
        mock_state = mock.MagicMock()
        mock_state.app = mock_app

        # Test fail through to default value
        record_state(mock_state)
        self.assertEqual(EVENTICS_CONFIG['EVENT_CALENDAR_PRODID'],
                         '-//CFPB//Event Calendar//EN')

        # Test using os.environ value
        with mock.patch.dict('os.environ', {'EVENT_CALENDAR_PRODID':
                             '-//CFPB//Another Event Calendar//EN'}):
            record_state(mock_state)
            self.assertEqual(EVENTICS_CONFIG['EVENT_CALENDAR_PRODID'],
                             '-//CFPB//Another Event Calendar//EN')

        # Test using the app.config value
        with mock.patch.dict(mock_state.app.config, {'EVENT_CALENDAR_PRODID':
                             '-//CFPB//Event Calendar//EN'}):
            record_state(mock_state)
            self.assertEqual(EVENTICS_CONFIG['EVENT_CALENDAR_PRODID'],
                             '-//CFPB//Event Calendar//EN')

    @mock.patch("requests.get")
    def test_get_event_json(self, mock_request_get):
        """
        Test that we're constructing the event source URL from which to fetch
        JSON correctly.
        """
        # We want two possible responses, first a good, 200 response, and
        # then a 404 response (a response that doesn't provide JSON). We
        # need to make sure we're handling the ValueError (JSONDecodeError).
        mock_good_response = mock.MagicMock()
        mock_good_response.status_code = 200
        mock_good_response.json.return_value = {'some': 'json'}

        mock_bad_response = mock.MagicMock()
        mock_bad_response.status_code = 404
        mock_bad_response.json.side_effect = ValueError()

        mock_request_get.side_effect = [
            mock_good_response,
            mock_bad_response
        ]

        EVENTICS_CONFIG['EVENT_SOURCE'] = 'http://localhost:9200/events/<event_slug>/'

        source_json, source_status = get_event_json('myevent')
        self.assertEqual(source_status, 200)
        mock_request_get.assert_called_with('http://localhost:9200/events/myevent/')

        source_json, source_status = get_event_json('myevent')
        self.assertEqual(source_status, 404)
        self.assertEqual(source_json, {})

    def test_generate_ics(self):
        """
        Test that, given a specific set of JSON, the generate_ics
        function generates valid iCalendar data.
        """
        event_json = {
            "ics": {
                "summary": "Test Event",
                "location": "Washington, DC",
                "uid": "8DB71F484FA2ABC57F621CB7F1@2013-07-03 09:30:00",
                "dtstart": "2015-07-01T05:00:00-04:00",
                'starting_tzinfo': 'America/New_York',
                "dtend": "2015-07-01T06:00:00-04:00",
                'ending_tzinfo': 'America/New_York',
                "dtstamp": "2013-07-02T14:29:08Z"
            }
        }

        with mock.patch('flask_eventics.controllers.get_event_json') as mock_get_event_json:
            mock_get_event_json.return_value = event_json, 200
            ics, status, headers = generate_ics('foo')

            # Make sure the ics parses
            try:
                icalendar.Calendar.from_ical(ics)
            except ValueError:
                self.fail("generate_ics() did not return a valid iCalendar file")

    def test_eventics(self):
        """
        Test that, as eventics registers and responds appropriately as a
        Flask Blueprint.
        """
        app = Flask(__name__)
        app.config['TESTING'] = True
        app.config['WTF_CSRF_ENABLED'] = False
        app.register_blueprint(eventics)

        client = app.test_client()
        response = client.get('/events/foo/ics')


if __name__ == '__main__':
    unittest.main()


flask_eventics/controllers.py
# -*- coding: utf-8 -*-

from werkzeug.routing import Rule, Map
from flask import Blueprint, request, Response, current_app

import os
from datetime import datetime
import dateutil.parser
import icalendar
from pytz import timezone

from urlparse import urlparse, urlunparse
import requests

eventics = Blueprint("flask-eventics", __name__, url_prefix="")

# Configuration Variables
# Other Flask-related projects at CFPB use environment variables rather than
# Flask configuration files. flask-eventics will provide the option for both,
# falling back on envvars if no configuration variables exist.
# Note: Flask doesn't make it easy for Blueprints to get access to app
# configuration. We have to capture that from the app state when the Blueprint
# is registered with the application.
EVENTICS_CONFIG = {
    # The URL pattern for ics files
    # MUST contain <event_slug> that corrosponds to the source lookup below
    'EVENT_ICS_URL': '/events/<event_slug>/ics',

    # A REST endpoint for events
    # Example: an elasticsearch endpoint
    'EVENT_SOURCE': 'http://localhost:9200/content/events/<event_slug>/_source',

    # Calendar product id
    'EVENT_CALENDAR_PRODID': '-//CFPB//Event Calendar//EN',

    # Field mapping for iCalendar fields from JSON object properties
    'EVENT_FIELD_SUMMARY': 'summary',
    'EVENT_FIELD_DTSTART': 'dtstart',
    'EVENT_FIELD_DTEND': 'dtend',
    'EVENT_FIELD_DTSTAMP': 'dtstamp',
    'EVENT_FIELD_UID': 'uid',
    'EVENT_FIELD_PRIORITY': 'priority',
    'EVENT_FIELD_ORGANIZER': 'organizer',
    'EVENT_FIELD_ORGANIZER_ADDR': 'organizer_email',
    'EVENT_FIELD_LOCATION': 'location',
    'EVENT_FIELD_STATUS': 'status',

    # Status is one of:
    # TENTATIVE/CONFIRMED/CANCELLED

    # Defaults
    'EVENT_DEFAULT_SUMMARY': '',
    'EVENT_DEFAULT_DTSTART': '2015-01-11T10:30:00Z',
    'EVENT_DEFAULT_DTEND': '2015-01-11T10:30:00Z',
    'EVENT_DEFAULT_DTSTAMP': '2015-01-11T10:30:00Z',
    'EVENT_DEFAULT_UID': '',
    'EVENT_DEFAULT_PRIORITY': 1,
    'EVENT_DEFAULT_ORGANIZER': '',
    'EVENT_DEFAULT_ORGANIZER_ADDR': '',
    'EVENT_DEFAULT_LOCATION': '',
    'EVENT_DEFAULT_STATUS': 'TENTATIVE',
}


def get_event_json(event_slug):
    """
    Fetch the event json from the configured event source
    (EVENTICS_CONFIG['EVENT_SOURCE']).

    Returns the JSON and a status code
    """
    # First, construct the url from EVENT_SOURCE, which uses Werkzeug URL
    # placeholders the same as our Flask url. This means constructing a simple
    # URL mapping.
    source_url_parts = urlparse(EVENTICS_CONFIG['EVENT_SOURCE'])
    source_map = Map([
        Rule(source_url_parts.path, endpoint="event_source")
    ])
    source_adapter = source_map.bind('')
    source_url_path = source_adapter.build('event_source', {'event_slug': event_slug})
    source_url = urlunparse(source_url_parts._replace(path=source_url_path))
    source_response = requests.get(source_url)

    try:
        return source_response.json(), source_response.status_code
    except ValueError:
        return {}, source_response.status_code


def generate_ics(event_slug):
    """
    Return an ics file for a given event.
    """

    # Get the event json from our source
    event_json, source_status = get_event_json(event_slug)
    if source_status != 200:
        return '', source_status

    # Create the Calendar
    calendar = icalendar.Calendar()
    calendar.add('prodid', EVENTICS_CONFIG['EVENT_CALENDAR_PRODID'])
    calendar.add('version', '2.0')
    calendar.add('method', 'publish')

    # Create the event
    event = icalendar.Event()

    # Populate the event
    ics_fields = event_json.get('ics')
    tz_dict = {'dtstart': 'starting_tzinfo', 'dtend': 'ending_tzinfo'}
    for ics_field, timezoneName in tz_dict.items():
        if ics_field in ics_fields:
            esdate = dateutil.parser.parse(ics_fields[ics_field])
            if timezoneName in ics_fields:
                date = esdate.astimezone(timezone(ics_fields[timezoneName]))
            else:
                date = esdate.astimezone(timezone('UTC'))
            event.add(ics_field, date)
            del ics_fields[ics_field]
            del ics_fields[timezoneName]
    for field in ics_fields:
        if field == 'dtstamp':
            esdate = dateutil.parser.parse(ics_fields[field])
            event.add(field, date)
        else:
            event.add(field, ics_fields[field])

    # Add the event to the calendar
    calendar.add_component(event)

    # Return the ICS formatted calendar
    return calendar.to_ical(), source_status, {
        'Content-Type': 'text/calendar; charset=utf-8',
        'Content-Disposition': 'attachment;filename=' + event_slug + '.ics'
    }


def record_state(state):
    """
    Record the application state when this blueprint is registered.
    Here we exact appropriate configuration values from the application
    config, if they're available, and use environment variables or
    defaults if not.
    """
    global EVENTICS_CONFIG
    app = state.app

    # For each of these values get the app config value. If it doesn't exist,
    # get an environment variable. If that doesn't exist, use the default.
    # This is a simple function to do that.
    get_config = lambda k: app.config.get(k, os.environ.get(k, EVENTICS_CONFIG[k]))

    for config_key in EVENTICS_CONFIG:
        EVENTICS_CONFIG[config_key] = get_config(config_key)

    # Set the route for generate_ics?
    eventics.route(EVENTICS_CONFIG['EVENT_ICS_URL'], methods=['GET'])(generate_ics)

# Weakness in Python decorators is that you can't mport and directly test the
# decorated functions, which is why this (and generate_ics) are
# undecorated.
eventics.record(record_state)
